### PR TITLE
ProcessGroupNCCL check tensor is_contiguous->is_non_overlapping_and_dense

### PR DESCRIPTION
**Pull request by zasdfgbnm**
I don't see any reason why we are asserting contiguous here. In my opinion, `is_non_overlapping_and_dense` makes more sense because this allows tasks like NHWC training.

Currently `check_gpu_tensors_same_device` and `check_gpu_tensors_different_devices` are both checking `is_non_overlapping_and_dense` instead of `is_contiguous`, so I think it makes more sense to modify `check_gpu_single_tensor` to have the same behavior

https://github.com/pytorch/pytorch/blob/f6c275f55ddfd22f8c0558efad3068b0e91f1560/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1290-L1292
https://github.com/pytorch/pytorch/blob/f6c275f55ddfd22f8c0558efad3068b0e91f1560/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1260-L1262

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -1221,8 +1221,8 @@ void check_gpu_single_tensor(const at::Tensor& tensor) {
   if (!tensor.is_cuda() || tensor.is_sparse()) {
     TORCH_CHECK(false, "Tensors must be CUDA and dense");
   }
-  if (!tensor.is_contiguous()) {
-    TORCH_CHECK(false, "Tensors must be contiguous");
+  if (!tensor.is_non_overlapping_and_dense()) {
```
Comment (kumpera) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1224]: Do we have tests covering this change?
We're now asking more of NCCL so we should cover this new condition with additional tests.

**Review by kumpera — CHANGES_REQUESTED** [review]
Shouldn't we add a tes

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1224]: I don't think the collectives that already support `is_non_overlapping_and_dense` have any test for this case, but can add tests for all of them in this PR.

**Review by kumpera — APPROVED** [review]
Thanks for the improvement!

**Comment by zasdfgbnm** [comment]
Some tests are failing

**Review by kumpera — CHANGES_REQUESTED** [review]
The test caused a regression in the GLOO tests:

======================================================================
ERROR [3.131s]: test_all_reduce_sum_complex (__main__.TestDistBackendWithSpawn)

The culprit is:

  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\distributed\distributed_test.py", line 435, in _build_tensor
    tensor.fill_(value)
RuntimeError: value cannot be converted to type float without overflow

**Pull request closed without merging**